_posts/2020-09-27-spring-boot-12-factor-app.md

```diff
@@ -0,0 +1,169 @@
+
+## Backing Services - Pluggable Datasources, and Queues
+> Treat backing services as attached resources
+
+Backing services should be attached and replaceable instead of embedded in the code. The use of specifications like JPA helps us achieve this for RDBMS databases. But in the absence of specifications, some code creeps into the code although we can keep them separate with abstraction layers. Similar to JPA, we can use JMS for messaging and SMTP for mails.
+```xml
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-data-jpa</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.h2database</groupId>
+      <artifactId>h2</artifactId>
+      <scope>runtime</scope>
+    </dependency>
+```
+The declarative format allows us to replace the H2 database with any other RDBMS like Oracle or MySQL.
+
+## Build, Release, Run - Leverage Development Workflow for Containers
+> Strictly separate build and run stages
+
+The stages for Build, Release, Run should be kept separate. For Spring Boot Applications, this is easy to achieve with the development workflow for containers. 
+
+The activities in these stages are:
+***Build***: we compile the source code and build the Docker Image
+***Release***: Tag the Image and push to the Registry.
+***Run***:  The Image pushed to the Registry during the release stage is pulled and run as a container instance. 
+ 
+If we are using containers to package and run our application, no application changes are required to adhere to the Twelve-factor app principle.
+
+
+## Processes - Stateless Applications
+> Execute the app as one or more stateless processes
+
+Spring Boot applications execute as a Java process on the host system or inside a container runtime environment like Docker. This principle advocates that the processes should be stateless and share-nothing.  Any data that needs to persist must be stored in a stateful backing service like a database.
+
+This is a shift from the method of using “sticky sessions” in web applications that cache user session data in the memory of the application's process and expecting future requests from the same session to be routed to the same process. Sticky sessions are a violation of twelve-factor. Session state data should be stored outside the application in a datastore that offers time-expiration, such as Memcached or Redis.
+
+## Port Binding - Port Defined as Environment Property
+> Export services via port binding
+
+Port binding is one of the fundamental requirements for microservices to be autonomous and self-contained.
```

Comment:
Please explain in a sentence what port binding is and why the 12 factor app thinks it's important.